azure-kit-starter/src/main/java/com/vaadin/azure/starter/ui/ClusterSupport.java

```diff
@@ -0,0 +1,245 @@
+        serviceInitEvent.addRequestHandler((RequestHandler) (vaadinSession,
+                vaadinRequest, vaadinResponse) -> {
+
+            // Set the thread local instance
+            CurrentInstance.set(ClusterSupport.class, this);
+
+            vaadinSession.access(() -> {
+
+                // Set always the version cookie
+                Cookie versionCookie = getCookieByName(CURRENT_VERSION_COOKIE);
+                if (versionCookie == null
+                        || !versionCookie.getValue().equals(appVersion)) {
+                    versionCookie = new Cookie(CURRENT_VERSION_COOKIE,
+                            appVersion);
+                    versionCookie.setHttpOnly(true);
+                    vaadinResponse.addCookie(versionCookie);
+                }
+
+                // Check always for a new version cookie
+                Cookie currentCookie = getCookieByName(UPDATE_VERSION_COOKIE);
+                if (currentCookie != null && !currentCookie.getValue().isEmpty()
+                        && !versionCookie.getValue()
+                                .equals(currentCookie.getValue())) {
+                    vaadinSession.getUIs().forEach(ui -> {
+                        if (ui.getChildren().anyMatch(
+                                c -> (c instanceof VersionNotificator))) {
+                            return;
+                        }
+                        VersionNotificator notificator = new VersionNotificator(
+                                appVersion, currentCookie.getValue());
+                        notificator.addSwitchVersionEvent(e -> {
+
+                            // Do nothing if switch version listener prevents
+                            // switching
+                            if (switchVersionListener != null
+                                    && !switchVersionListener.clusterSwitch(
+                                            VaadinRequest.getCurrent(),
+                                            VaadinResponse.getCurrent())) {
+                                return;
+                            }
+                            // When the users click on the notificator remove
+                            // cluster and session cookies
+                            removeCookie(STICKY_CLUSTER_COOKIE);
+                            removeCookie(CURRENT_VERSION_COOKIE);
+                            removeCookie(UPDATE_VERSION_COOKIE);
+                            // Invalidate session, vaadin does not synchronizes
+                            // it between clusters
+                            VaadinRequest.getCurrent().getWrappedSession()
+                                    .invalidate();
+                        });
+
+                        // Show notificator
+                        ui.add(notificator);
+                    });
+                }
+
+            });
+
+            // If current and new version does not match, notify user
+            return false;
+        });
+    }
```

Comment:
Let's extract all these nested lambdas to private methods to improve readability.

Reply:
Also, removed some unnecessary code around getting the app version.